tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc

```diff
@@ -5445,97 +5578,84 @@ Status ConvertFullyConnectedHelper(OpConverterParams* params,
-    return ConvertFullyConnectedHelper(
-        params, input_a.tensor(), input_b.weights(), transpose_b, node_def);
-  }
-
-  const auto get_matrix_op = [](nvinfer1::ITensor* in,
-                                bool transpose) -> nvinfer1::MatrixOperation {
-    return (in->getDimensions().nbDims < 2) ? nvinfer1::MatrixOperation::kVECTOR
-           : (transpose) ? nvinfer1::MatrixOperation::kTRANSPOSE
-                         : nvinfer1::MatrixOperation::kNONE;
-  };
+  bool converted = false;
+  TF_RETURN_IF_ERROR(ConvertFullyConnectedHelper(params, input_a, input_b,
+                                                 transpose_a, transpose_b,
+                                                 node_def, &converted));
+  if (converted == true) return Status::OK();
 
-  // If the MatMul operand is a constant, applies transposes at conversion-time
-  // as necessary. If the operand is a tensor, does nothing. If required
-  // transposes were applied, sets transpose to false.
+  // Add constant layer for weight operand, otherwise do nothing.
   const auto prepare_matmul_operand =
       [&params](TRT_TensorOrWeights operand,
                 bool* transpose) -> nvinfer1::ITensor* {
     if (operand.is_tensor()) {
       return operand.tensor();
     } else {
-      TRT_ShapedWeights weights(operand.weights().TrtDType());
-      if (*transpose) {
-        weights = params->weight_store->GetTempWeights(operand.weights());
-        ReorderCKtoKC(operand.weights(), &weights);
-        // Weights have been transposed, can set transpose to false
-        *transpose = false;
-      } else {
-        weights = operand.weights();
-      }
-      return params->converter->CreateConstantLayer(weights, weights.shape_);
+      return params->converter->CreateConstantLayer(operand.weights(),
+                                                    operand.GetTrtDims());
     }
   };
 
   nvinfer1::ITensor* tensor_a = prepare_matmul_operand(input_a, &transpose_a);
   nvinfer1::ITensor* tensor_b = prepare_matmul_operand(input_b, &transpose_b);
 
+  const auto get_matrix_op = [](nvinfer1::ITensor* in,
+                                bool transpose) -> nvinfer1::MatrixOperation {
+    return (transpose) ? nvinfer1::MatrixOperation::kTRANSPOSE
+                       : nvinfer1::MatrixOperation::kNONE;
+  };
+  nvinfer1::MatrixOperation op_a, op_b;
+  // Note: in implicit batch mode kTRANSPOSE and kNONE is only valid if the
+  // matrix has at least 2 non batch dimension. In implicit batch mode, if a has
+  // 1 dim (excluding batch dim), then we can treat it as a batch of vectors.
+  op_a = (tensor_a->getDimensions().nbDims < 2)
+             ? nvinfer1::MatrixOperation::kVECTOR
+             : get_matrix_op(tensor_a, transpose_a);
+  // In implicit batch mode, if B has only 1 dims (excluding batch dim) then we
+  // cannot convert. The kVECTOR trick is not applicable for B, because the
+  // output would not have correct layout.
```

Comment:
Let's modify the comment:
 In implicit batch mode, we already reject the case where tensor_b is 1D, as TensorRT doesn't allow the use of kVECTOR on this operand.